src/node/interfaces.cpp

```diff
@@ -838,6 +839,52 @@ class ChainImpl : public Chain
     NodeContext& m_node;
 };
 
+class BlockTemplateImpl : public BlockTemplate
+{
+public:
+    explicit BlockTemplateImpl(std::unique_ptr<CBlockTemplate> block_template) : m_block_template(std::move(block_template)) {
+        m_block_header = m_block_template->block.GetBlockHeader();
+    }
+
+    std::unique_ptr<CBlockTemplate> m_block_template;
+    CBlockHeader m_block_header;
```

Comment:
In commit "Have createNewBlock return BlockTemplate interface" (b4782696a8ec5e21a41db3702d4b4dcc90ea40dd)

Could you move these member declarations to the bottom of the class? It makes it easier to see what state the interface accesses when state is declared at beginning or end of the class, not in the middle mixed with method definitions. 

I'm also wondering if it might be more efficient to avoid the call to CBlock::GetBlockHeader():

```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -843,14 +843,10 @@ class BlockTemplateImpl : public BlockTemplate
 {
 public:
     explicit BlockTemplateImpl(std::unique_ptr<CBlockTemplate> block_template) : m_block_template(std::move(block_template)) {
-        m_block_header = m_block_template->block.GetBlockHeader();
     }
 
-    std::unique_ptr<CBlockTemplate> m_block_template;
-    CBlockHeader m_block_header;
-
     CBlockHeader getBlockHeader() override {
-        return m_block_header;
+        return Assert(m_block_template)->block;
     }
 
     CBlock getBlock() override
@@ -883,6 +879,7 @@ public:
         return GetWitnessCommitmentIndex(m_block_template->block);
     }
 
+    std::unique_ptr<CBlockTemplate> m_block_template;
 };
 
 class MinerImpl : public Mining
```

Reply:
Do you mean return `Assert(m_block_template)->block.header`?

The reason I made it a member was so that the interface would return it upon constructing `BlockTemplate`. This would then save multiple round trips over the interface later when we need a bunch of details from the header.

Reply:
> Do you mean return `Assert(m_block_template)->block.GetBlockHeader()`?

No I think the diff above should work. There should be no need for calling GetBlockHeader() at all, much less caching the result of the call, because CBlock inherits from CBlockHeader.

I'm not suggesting dropping the getBlockHeader interface method, because that method is useful for avoiding round trips like you suggest.

Reply:
> because CBlock inherits from CBlockHeader

Ah ok

> that method is useful for avoiding round trips

The choice is between having the caller be careful to avoid round trips, vs not needing round trips at all because we already got the header at construction time.

That's likely a tiny difference, so whatever is better code is probably best.

Reply:
I took the patch.